src/Nethermind/Nethermind.Runner/scripts/app.ts

```diff
@@ -0,0 +1,400 @@
+
+  // Update the D3 treemap with the merged data
+  updateTreemap<TransactionReceipt>(
+    document.getElementById("block"), // or d3.select("#treemap")
+    200,
+    parseInt(data.head.gasLimit, 16),
+      mergedData,
+      // keyFn
+      d => d.hash,
+      // orderFn
+      d => d.order,
+      // sizeFn
+      d => parseInt(d.gasUsed, 16),
+      // colorFn
+      d => parseInt(d.effectiveGasPrice, 16) * parseInt(d.gasUsed, 16)
+  );
+
+  // Update the rolling box plot for this block
+  // passing the blockNum as well:
+  boxPlotEGP.update(mergedData, number);
+});
+
+sse.addEventListener("peers", (e) => {
+  const data = JSON.parse(e.data) as Peer[];
+
+  // e.data is a JSON array of Peer objects
+  const peersArray = JSON.parse(e.data) as Peer[];
+
+  // Aggregate by clientType
+  const countsByType: Record<string, number> = {};
+
+  peersArray.forEach((peer) => {
+    const nodeType = getNodeType(peer.clientType);
+    countsByType[nodeType] = (countsByType[nodeType] || 0) + 1;
+  });
+
+  // Convert to an array for D3
+  const peers = Object.entries(countsByType).map(([type, count]) => ({
+    type,
+    count
+  }));
+
+  // Now update the chart
+  updatePieChart(peers);
+});
+
+let maxCpuPercent = 0;
+let maxMemoryMb = 0;
+sse.addEventListener("system", (e) => {
+  const data = JSON.parse(e.data) as System;
+  let memoryMb = data.workingSet / (1024 * 1024);
+  if (memoryMb > maxMemoryMb) {
+    maxMemoryMb = memoryMb;
+  }
+  let cpuPercent = (data.userPercent + data.privilegedPercent) * 100;
+  if (cpuPercent > maxCpuPercent) {
+    maxCpuPercent = cpuPercent;
+  }
+
+  const now = performance.now();
+  addCapped(seriesTotalCpu, { t: now, v: data.userPercent + data.privilegedPercent });
+  addCapped(seriesMemory, { t: now, v: memoryMb });
+
+  if (document.hidden) return;
+
+  updateText(upTime, formatDuration(data.uptime));
+
+  updateText(cpuTime, formatDec(cpuPercent));
+  updateText(maxCpuTime, formatDec(maxCpuPercent));
+  sparkline(sparkCpu, seriesTotalCpu, 150, 100, 60);
+
+  updateText(memory, format(memoryMb));
+  updateText(maxMemory, format(maxMemoryMb));
+  sparkline(sparkMemory, seriesMemory, 150, 100, 60);
+});
+
+let logs: string[] = [];
+sse.addEventListener("log", (e) => {
+  const data = JSON.parse(e.data) as string[];
+  for (let entry of data) {
+    const html = ansiConvert.toHtml(entry);
+    if (logs.length > 100) { logs.shift(); }
+    logs.push(html);
+  }
+});
+
+function appendLogs() {
+  requestAnimationFrame(appendLogs);
+  if (logs.length > 0) {
+    let scroll = false;
+    if (nodeLog.scrollHeight < 500 || nodeLog.scrollTop < nodeLog.scrollHeight - 500) {
+      scroll = true;
+    }
+    const frag = document.createDocumentFragment();
+    for (let i = 0; i < logs.length; i++) {
+      const newEntry = document.createElement('div');
+      newEntry.innerHTML = logs[i];
+      frag.appendChild(newEntry);
+    }
+    logs = [];
+    nodeLog.appendChild(frag);
```

Comment:
Do we drain this `nodeLog` element at some point? Otherwise if we keep appending then we'll run OOM.